bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java

```diff
@@ -739,136 +743,177 @@ private void verifyTaskIsRegistered(PollTask task) throws PollTaskUnregistered {
+        public ScheduledFuture<?> submitOneTimePoll(ModbusReadRequestBlueprint request,
+                @Nullable ModbusReadCallback callback) {
+            if (closed) {
+                throw new IllegalStateException("Communication interface is closed already!");
+            }
+            ScheduledExecutorService executor = scheduledThreadPoolExecutor;
             Objects.requireNonNull(executor, "Not activated!");
-            Objects.requireNonNull(factory, "Not activated!");
+            long scheduleTime = System.currentTimeMillis();
+            BasicPollTask task = new BasicPollTask(endpoint, request, callback);
+            logger.debug("Scheduling one-off poll task {}", task);
+            ScheduledFuture<?> future = executor.schedule(() -> {
+                long millisInThreadPoolWaiting = System.currentTimeMillis() - scheduleTime;
+                logger.debug("Will now execute one-off poll task {}, waited in thread pool for {}", task,
+                        millisInThreadPoolWaiting);
+                executeOperation(task, true, pollOperation);
+            }, 0L, TimeUnit.MILLISECONDS);
+            return future;
+        }
 
-            // cancel poller
-            @Nullable
-            ScheduledFuture<?> future = scheduledPollTasks.remove(task);
-            if (future == null) {
-                // No such poll task
-                logger.warn("Caller tried to unregister nonexisting poll task {}", task);
-                return false;
+        @Override
+        public PollTask registerRegularPoll(ModbusReadRequestBlueprint request, long pollPeriodMillis,
+                long initialDelayMillis, @Nullable ModbusReadCallback callback) {
+            synchronized (ModbusManagerImpl.this) {
+                if (closed) {
+                    throw new IllegalStateException("Communication interface is closed already!");
+                }
+                ScheduledExecutorService executor = scheduledThreadPoolExecutor;
+                Objects.requireNonNull(executor, "Not activated!");
+                BasicPollTask task = new BasicPollTask(endpoint, request, callback);
+                logger.trace("Registering poll task {} with period {} using initial delay {}", task, pollPeriodMillis,
+                        initialDelayMillis);
+                if (scheduledPollTasks.containsKey(task)) {
+                    logger.trace("Unregistering previous poll task (possibly with different period)");
+                    unregisterRegularPoll(task);
+                }
+                ScheduledFuture<?> future = executor.scheduleWithFixedDelay(() -> {
+                    long started = System.currentTimeMillis();
+                    logger.debug("Executing scheduled ({}ms) poll task {}. Current millis: {}", pollPeriodMillis, task,
+                            started);
+                    try {
+                        executeOperation(task, false, pollOperation);
+                    } catch (RuntimeException e) {
```

Comment:
Cherry-picked 5bde9d6604375f1212789dae5e002c5b7e9ceff3 on rebase in this commit.